packages/state-importer/src/main.js

```diff
@@ -0,0 +1,240 @@
+    if (!options.apiConfig)
+      throw new StateImportError(
+        'The constructor must be passed an `apiConfig` object'
+      )
+
+    this.apiConfig = options.apiConfig
+    this.accessToken = options.accessToken
+    this.continueOnProblems = options.continueOnProblems || false
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          this.accessToken ? `Bearer ${this.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    this.syncStates = createSyncStates()
+
+    this.logger = logger || {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      debug: () => {},
+    }
+    this._resetSummary()
+  }
+
+  static _buildPredicate(states: Array<StateData>) {
+    return `key in ("${states.map(state => state.key).join('", "')}")`
+  }
+
+  // Wrapper function for compatibility with CLI
+  processStream(chunk: Array<StateData>, cb: () => mixed) {
+    this.logger.debug(`Starting conversion of ${chunk.length} states`)
+    return this._processBatches(chunk).then(cb)
+    // No catch block as errors will be caught in the CLI
+  }
+
+  // Public function for direct module usage
+  run(states: Array<StateData>) {
+    return this._processBatches(states)
+  }
+
+  _processBatches(states: Array<StateData>) {
+    const predicate = StateImport._buildPredicate(states)
+    const service = this._createService()
+    const uri = service.where(predicate).build()
+    const existingStatesRequest = { uri, method: 'GET' }
+    return this.client
+      .execute(existingStatesRequest)
+      .then(({ body: { results: existingStates } }: Object) =>
+        this._createOrUpdate(states, existingStates)
+      )
+      .then(() => Promise.resolve())
+      .catch(error => {
+        // format error and throw to CLI error handler
+        throw new StateImportError(
+          'Processing batch failed',
+          error.message || error,
+          this._summary
+        )
+      })
+  }
+
+  _createOrUpdate(
+    newStates: Array<StateData>,
+    existingStates: Array<StateData>
+  ) {
+    return Promise.map(newStates, newState => {
+      const existingState = existingStates.find(
+        state => state.key === newState.key
+      )
+      if (existingState)
+        return this._update(newState, existingState)
+          .then(response => {
+            if (response && response.statusCode === 304)
+              this._summary.unchanged += 1
+            else this._summary.updated += 1
+            return Promise.resolve()
+          })
+          .catch(error => {
+            if (this.continueOnProblems) {
+              this._summary.updateErrorCount += 1
+              this._summary.errors.push(error.message || error)
```

Comment:
Why do we need to keep track of both? Count and an error in itself. Feels like wonky bookkeeping. One is an derived value of the other. Either `errors.updates` to which we push for a `errors.filter(error => error.isUpdateError)` would do the same?